replay/models/cql.py

```diff
@@ -0,0 +1,357 @@
+        self.action_randomization_scale = action_randomization_scale
+        self.n_epochs = n_epochs
+        self.assert_omp_single_thread()
+
+        self.model = CQL_d3rlpy.CQL(
+            actor_learning_rate=actor_learning_rate,
+            critic_learning_rate=critic_learning_rate,
+            temp_learning_rate=temp_learning_rate,
+            alpha_learning_rate=alpha_learning_rate,
+            actor_optim_factory=actor_optim_factory,
+            critic_optim_factory=critic_optim_factory,
+            temp_optim_factory=temp_optim_factory,
+            alpha_optim_factory=alpha_optim_factory,
+            actor_encoder_factory=actor_encoder_factory,
+            critic_encoder_factory=critic_encoder_factory,
+            q_func_factory=q_func_factory,
+            batch_size=batch_size,
+            n_frames=n_frames,
+            n_steps=n_steps,
+            gamma=gamma,
+            tau=tau,
+            n_critics=n_critics,
+            initial_temperature=initial_temperature,
+            initial_alpha=initial_alpha,
+            alpha_threshold=alpha_threshold,
+            conservative_weight=conservative_weight,
+            n_action_samples=n_action_samples,
+            soft_q_backup=soft_q_backup,
+            use_gpu=use_gpu,
+            scaler=scaler,
+            action_scaler=action_scaler,
+            reward_scaler=reward_scaler,
+            **params
+        )
+
+    def _fit(
+        self,
+        log: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> None:
+        train: MDPDataset = self._prepare_data(log)
+        self.model.fit(train, n_epochs=self.n_epochs)
+
+    def _prepare_data(self, log: DataFrame) -> MDPDataset:
+        user_logs = log.toPandas().sort_values(['user_idx', 'timestamp'], ascending=True)
+
+        # reward top-K watched movies with 1, the others - with 0
+        user_top_k_idxs = (
+            user_logs
+            .sort_values(['relevance', 'timestamp'], ascending=False)
+            .groupby('user_idx')
+            .head(self.top_k)
+            .index
+        )
+        rewards = np.zeros(len(user_logs))
+        rewards[user_top_k_idxs] = 1.0
+
+        # every user has his own episode (the latest item is defined as terminal)
+        user_terminal_idxs = (
+            user_logs[::-1]
+            .groupby('user_idx')
+            .head(1)
+            .index
+        )
+        terminals = np.zeros(len(user_logs))
+        terminals[user_terminal_idxs] = 1
+
+        # cannot set zero scale as d3rlpy will treat transitions then as discrete
+        assert self.action_randomization_scale > 0
+        action_randomization_scale = self.action_randomization_scale
+        action_randomization = np.random.randn(len(user_logs)) * action_randomization_scale
+
+        actions = user_logs['relevance'].to_numpy().astype(float) + action_randomization
+
+        train_dataset = MDPDataset(
```

Comment:
you can do first part on Spark and only here start to use Pandas

Reply:
Well, it feels in the spirit of `RePlay` to use Spark here. However, I still see more reasons to make it non-distributedly, because:

This process is relatively cheap — traverse, sorting, grouping, no costly joins. The resulting data in terms of size is equal to the input data. Using distributed calculations in my opinion will only degrade performance for the vast majority of the input size range while also complicating the code itself.

I understand my points are somewhat speculative, but still.

Reply:
it is may be true for small datasets, but not for large datasets.

Could you tell me, why the fit time is so large?